lib/modules/manager/gradle/extract/consistent-versions-plugin.ts

```diff
@@ -0,0 +1,187 @@
+}
+
+/**
+ * Parses Gradle-Consistent-Versions files to figure out what dependencies, versions
+ * and groups they contain. The parsing goes like this:
+ * - Parse `versions.props` into deps (or groups) and versions, remembering file offsets
+ * - Parse `versions.lock` into deps and lock-versions
+ * - For each exact dep in props file, lookup the lock-version from lock file
+ * - For each group/regex dep in props file, lookup the set of exact deps and versions in lock file
+ *
+ * @param propsFileContent text content of `versions.props
+ * @param lockFileContent text content of `versions.lock`
+ */
+export function parseGcv(
+  propsFileContent: string,
+  lockFileContent: string
+): PackageDependency<GradleManagerData>[] {
+  const lockFileMap = parseLockFile(lockFileContent);
+  const [propsFileExactMap, propsFileRegexMap] =
+    parsePropsFile(propsFileContent);
+
+  const extractedDeps: PackageDependency<GradleManagerData>[] = [];
+
+  // For each exact dep in props file
+  for (const [propDep, versionAndPosition] of propsFileExactMap) {
+    if (lockFileMap.has(propDep)) {
+      const newDep: Record<string, any> = {
+        managerData: {
+          packageFile: VERSIONS_PROPS,
+          fileReplacePosition: versionAndPosition.filePos,
+        },
+        packageName: propDep,
+        currentValue: versionAndPosition.version,
+        currentVersion: versionAndPosition.version,
+        lockedVersion: lockFileMap.get(propDep),
+      } as PackageDependency<GradleManagerData>;
+      extractedDeps.push(newDep);
+    }
+  }
+
+  // For each regular expression (group) dep in props file
+  for (const [propDepRegEx, propVerAndPos] of propsFileRegexMap) {
+    for (const [exactDep, exactVer] of lockFileMap) {
+      if (propDepRegEx.test(exactDep)) {
+        const newDep: Record<string, any> = {
+          managerData: {
+            packageFile: VERSIONS_PROPS,
+            fileReplacePosition: propVerAndPos.filePos,
+          },
+          depName: exactDep,
+          currentValue: propVerAndPos.version,
+          currentVersion: propVerAndPos.version,
+          lockedVersion: exactVer,
+          groupName: regexToGroupString(propDepRegEx),
+        } as PackageDependency<GradleManagerData>;
+        extractedDeps.push(newDep);
+      }
+    }
+  }
+  return extractedDeps;
+}
+
+//----------------------------------
+// Private utility functions below
+//----------------------------------
+
+// Translate glob syntax to a regex that does the same
+function globToRegex(depName: string): RegExp {
+  return regEx(
+    depName
+      .replaceAll('*', '_WC_CHAR_')
+      .replace(/[/\-\\^$*+?.()|[\]{}]/g, '\\$&')
+      .replaceAll('_WC_CHAR_', '[^:]*')
+  );
+}
+
+// Translate the regex from versions.props into a group name used in branch names etc
+function regexToGroupString(regExp: RegExp): string {
+  return regExp.source.replaceAll('[^:]*', '-').replaceAll('\\.', '.');
+}
```

Comment:
Are you sure there isn't a smarter way to solve to achieve this than to dynamically construct RegEx patterns, in which you then replace individual characters back and forth?

Shouldn't the groupId, e.g. `org.apache.lucene` in `version.props` already be good as a group name? 
Or is there a reason to specifically build something like `groupName: 'org.apache.lucene:-'`?

Reply:
I believe the plugin supports glob asterisk anywhere in the depName, and you could theoretically have both `org.apache.lucene:*=1`, `org.apache.lucene:foo*=2` and `org.apache.lucene:foo-bar=3` in the same file. So I guess I tried to let renovate construct a branch for `org.apache.lucene:foo-` as well as `org.apache.lucene:-`. 

But what you're saying is that if we could set the group to the maven group, so that all `org.apachee.lucene` changes are handled in the same git branch/PR, even though they may touch two different props-file lines due to the globs defined in the props file.

I think it could work, as long as you can assume that maven groupId is different for unrelated artifacts from the same vendor. If that is what the rest of gradle manager does, then we should be fine!

Reply:
Currently, gradle manager sets `groupName` only if versions are set in variables, i.e., multiple deps could refer to that. I guess always taking the maven group won't be an option if `*` can be placed everywhere. 

A quick search just revealed `com.fasterxml.jackson.*:* = 2.11.4` [here](https://github.com/apache/iceberg/blob/49a0ea956e3a4b9979754c887d803d4ab51131ae/versions.props#L10). I'm not familiar with the GCV plugin at all but just stumbled upon [FuzzyPatternResolver.java](https://github.com/palantir/gradle-consistent-versions/blob/develop/src/main/java/com/palantir/gradle/versions/FuzzyPatternResolver.java#L104) -> seems to me they're also constructing RegEx patterns dynamically.

Reply:
Nice find re the source code of palantir's glob code. In Java you can compile a regex, and I see they cache the compiled ones for more speed. Not sure if you can do similar in TS.

Also interesting to see that they replace `*` with `.*?` regex, i.e. the pattern you found `com.fasterxml.jackson.*:*` would be equivalent with `com.fasterxml.jackson.*`. I'll change my regex to do the same, and not stop at `:`.